src/lib.rs

```diff
@@ -201,10 +201,13 @@ pub async fn development_transport(
     keypair: identity::Keypair,
 ) -> std::io::Result<core::transport::Boxed<(PeerId, core::muxing::StreamMuxerBox)>> {
     let transport = {
-        let tcp = tcp::TcpConfig::new().nodelay(true);
-        let dns_tcp = dns::DnsConfig::system(tcp).await?;
-        let ws_dns_tcp = websocket::WsConfig::new(dns_tcp.clone());
-        dns_tcp.or_transport(ws_dns_tcp)
+        let dns_tcp = || {
+            let tcp_config = tcp::GenTcpConfig::default().nodelay(true);
+            let tcp = tcp::TcpTransport::new(tcp_config);
+            dns::DnsConfig::system(tcp)
+        };
+        let ws_dns_tcp = websocket::WsConfig::new(dns_tcp().await?);
+        dns_tcp().await?.or_transport(ws_dns_tcp)
```

Comment:
Opening a discussion thread here related to what @thomaseizinger wrote in https://github.com/libp2p/rust-libp2p/pull/2652#discussion_r884856191
Since the original discussion was about the ListenerId I am moving this discussion here.

> > Yes we clone "leaf" transports like tcp. For example in development_transport, we create two tcp transports wrapped in a dns transport, and then wrap one in a WsConfig and combine them in a OrTransport. If we would only bubble up the ids generated by the tcp transport, both "or_transport"-ed transports would generate the same Ids:
>
> Looking at the implementation of DnsConfig, the inner transport is Arc<Mutex>'d, meaning there is only one inner tcp transport, regardless how many times the wrapped DNS transport is cloned. Consequently, there is only one source of ListenerIds and there should never be a collision.

Within this PR I am removing the Clone derivation on `DnsConfig` and `WsConfig`, so with the code here I would actually not clone a single `Mutex`'ed TCP transport but instead create two separate TCP Transports. See the above implementation for construction two dns_tcp transports.
I think we have to do it this way because otherwise the race condition of polling the inner transport will cause errors.
For example in WsConfig:
We wrap an tcp transport, and for all addresses that have a `/ws` or `/wss` suffix we apply some logic, more specifically we apply logic on the `TransportEvents` associated with `ws` or `wss` listeners. In particular, we map the inbound upgrade on `TransportEvent::Incoming`. Now the problem is that because we are "second" in an `OrTransport`, any event that is happening on the inner TCP Transport will never make it to use. Because if the waker is polled, the "first" transport, that is the non-websocket one, will be polled first, poll the inner tcp transport and get the event.

@mxinden @thomaseizinger because of this, I think we can never allow Clone on a transport in general, and thus users will have to build the same transport multiple times. Please correct me in case that I am missing something.

Reply:
Yep, I think you are correct. It also feels more natural to _not_ clone `Transport`s, now that they have `&mut` receivers.